realm/realm-library/src/androidTest/kotlin/io/realm/MixedChangeListenerTests.kt

```diff
@@ -0,0 +1,485 @@
+                    MixedType.DECIMAL128 -> list.add(arrayOf(
+                            MixedType.DECIMAL128,
+                            Mixed.valueOf(Decimal128(1)),
+                            Mixed.valueOf(Decimal128(10)),
+                            Mixed.valueOf(10.5.toFloat()))
+                    )
+                    MixedType.OBJECT_ID -> list.add(arrayOf(
+                            MixedType.OBJECT_ID,
+                            Mixed.valueOf(ObjectId(Date(10))),
+                            Mixed.valueOf(ObjectId(Date(100))),
+                            Mixed.valueOf(Date(100)))
+                    )
+                    MixedType.UUID -> list.add(arrayOf(
+                            MixedType.UUID,
+                            Mixed.valueOf(UUID.randomUUID()),
+                            Mixed.valueOf(UUID.randomUUID()),
+                            Mixed.valueOf("hello world1"))
+                    )
+                    MixedType.OBJECT -> {
+                        val first = AllJavaTypes(0)
+                        first.fieldString = "FOO"
+
+                        val second = AllJavaTypes(0)
+                        second.fieldString = "FOO"
+                        second.fieldMixed = Mixed.valueOf(first)
+
+                        list.add(arrayOf(
+                                MixedType.OBJECT,
+                                first,
+                                second,
+                                "hello world1")
+                        )
+                    }
+                    MixedType.NULL -> { // Not tested directly
+                    }
+
+                    else -> throw AssertionError("Missing case for type: ${type.name}")
+                }
+            }
+
+            return list
+        }
+    }
+
+    @Test
+    fun primitives_changeValueKeepType() {
+        // Do not test on object types
+        if (testingType == MixedType.OBJECT)
+            return
+
+        looperThread.runBlocking {
+            Realm.getInstance(realmConfiguration).use { realm ->
+                realm.beginTransaction()
+
+                val all = AllJavaTypes(0)
+                all.fieldString = "FOO1"
+                all.fieldMixed = first as Mixed?
+                val managedAll = realm.copyToRealm(all)
+
+                realm.commitTransaction()
+
+                looperThread.keepStrongReference(managedAll)
+
+                RealmObject.addChangeListener<RealmModel>(managedAll, { _, changeSet ->
+                    changeSet!!
+                    looperThread.postRunnable(Runnable {
+                        Assert.assertTrue(changeSet.isFieldChanged(AllJavaTypes.FIELD_STRING))
+                        Assert.assertTrue(changeSet.isFieldChanged(AllJavaTypes.FIELD_MIXED))
+                        looperThread.testComplete()
+                    })
+                })
+
+                realm.beginTransaction()
+
+                val updatingModel = AllJavaTypes(0)
```

Comment:
Why not directly modifying `managedAll`? It would make the test short and easier to read?